### DISCUSSION TITLE
[macOS] Running FlareSolverr as a service (launchd agent)

### DISCUSSION THREAD

**Discussion by XxUnkn0wnxX** [FlareSolverr/FlareSolverr#99] [General]
Hi thought I'd share my implementation of running `FlareSolverr`

be sure to have node installed via homebrew: https://brew.sh
be sure you have the dir where homebrew links all its Formula in `/etc/paths`

mine looks like this, I've added `/usr/local/bin` to the top.
```
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
```

to find out where brew symlinks all its `Formula` run `brew --prefix node` in terminal & it should output the full path to node as well as the folder where everything else is linked, for me its `/usr/local/bin` & M1 macs will typically use `/opt/homebrew/bin`

now either using https://www.soma-zone.com/LaunchControl/ or you just modify the following plsit to suit your environment.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>EnvironmentVariables</key>
	<dict>
		<key>CAPTCHA_SOLVER</key>
		<string>none</string>
		<key>LOG_HTML</key>
		<string>false</string>
		<key>LOG_LEVEL</key>
		<string>info</string>
		<key>PATH</key>
		<string>/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin</string>
		<key>PORT</key>
		<string>8085</string>
	</dict>
	<key>KeepAlive</key>
	<true/>
	<key>Label</key>
	<string>FlareSolverr</string>
	<key>ProcessType</key>
	<string>Interactive</string>
	<key>ProgramArguments</key>
	<array>
		<string>/usr/local/bin/node</string>
		<string>dist/index.js</string>
	</array>
	<key>RunAtLoad</key>
	<true/>
	<key>StandardErrorPath</key>
	<string>logs/FlareSolverr-error.log</string>
	<key>StandardOutPath</key>
	<string>logs/FlareSolverr.log</string>
	<key>WorkingDirectory</key>
	<string>/Users/USER/Apps/FlareSolverr</string>
</dict>
</plist>
```
save as `FlareSolverr.plist`  in `~/Library/LaunchAgents` folder, type `open ~/Library/LaunchAgents` in terminal & it should open the folder.

![Screen Shot 2021-03-19 at 3 56 06 am](https://user-images.githubusercontent.com/2242880/111665836-3c52cb00-8867-11eb-9f05-5f59f5f319f0.jpg)

where `<string>8085</string>` change the port to whatever port value you need, 
where `<string>/Users/USER/Apps/FlareSolverr</string>` `USER` this will be your user folder name, or just copy & paste the full path to where you git cloned the "FlareSolverr" folder to

where `<string>/usr/local/bin/node</string>` this is the full path location to where node is installed via homebrew (once homebrew has installed node) just type in `brew --prefix node` in terminal & it should output the full path.
this location will be different on ARM mac's.

where
```
	<key>RunAtLoad</key>
	<true/>
```

set this to false if you don't want it to run on boot.

make the directory `logs` In the root of the `FlareSolverr` folder so logging works.
![Screen Shot 2021-03-19 at 3 51 44 am](https://user-images.githubusercontent.com/2242880/111665451-de25e800-8866-11eb-852f-fd9012b271b9.jpg)

every time you start ur mac it should autorun on load, to load & unload use: 

To Load:
`launchctl load ~/Library/LaunchAgents/FlareSolverr.plist`

To Unload: <-- be sure to use this before doing any git pull requests to update `FlareSolverr`
`launchctl unload ~/Library/LaunchAgents/FlareSolverr.plist`

**Comment by ghost** [comment]
having some trouble with my version of your .plist

```
Mon Aug 23 12:28:30
iMac191:~ john$ brew --prefix node
/usr/local/opt/node
```

<img width="212" alt="screenshot_ 2021-08-23 at 1 09 24 PM" src="https://user-images.githubusercontent.com/9438406/130495951-de31cccc-880e-4441-a97b-cedd8200cfa5.png">

EDIT:
I  used the path for the brew alias inside of `usr/local/bin` instead of the path given by the brew command you mention, `brew --prefix node`

**Comment by ngosang** [comment]
[FlareSolverr v2.0.0](https://github.com/FlareSolverr/FlareSolverr/releases/tag/v2.0.0) is out.

**Comment by ghost** [comment]
@XxUnkn0wnxX

Can't find the `index.js`  that I need to link to after version 2.0 release. 

```
Error: Cannot find module '/Users/john/Documents/GitHub/FlareSolverr/dist/index.js'
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:79:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

EDIT:
replace `dist/index.js` with `dist/server.js`

**Comment by ghost** [comment]
@XxUnkn0wnxX Any updates on running FlareSolverr 3.0 as a service on macOS. I can get it to run from Terminal but can't seem to construct a plist that works.

I have

**Comment by MrHoofer** [comment]
@XxUnkn0wnxX hey could you add me on discord please? hoping you can help me with a question Hoof#0900
